pubsub/redis.go

```diff
@@ -0,0 +1,224 @@
+//go:build events_redis
+
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/boz/go-lifecycle"
+	"github.com/go-redis/redis/v8"
+)
+
+type bus struct {
+	ctx    context.Context
+	client *redis.Client
+}
+
+type subscriber struct {
+	ctx           context.Context
+	cancel        func()
+	client        *redis.Client
+	lc            lifecycle.Lifecycle
+	eventch       chan Event
+	subscriptions map[string]bool
+}
+
+var _ Bus = (*bus)(nil)
+
+var _ Subscriber = (*subscriber)(nil)
+
+func NewBus() Bus {
+	b := &bus{
+		ctx: context.Background(),
+	}
+
+	b.client = redis.NewClient(&redis.Options{
+		Addr: getRedisHost(),
+	})
+
+	return b
+}
+
+func getRedisHost() string {
+	if value, ok := os.LookupEnv("AKASH_EVENTBUS_REDIS"); ok {
+		return value
+	}
+	return "localhost:6379"
+}
+
+func newSubscriber(topics ...string) *subscriber {
+	b := &subscriber{
+		lc:            lifecycle.New(),
+		eventch:       make(chan Event),
+		subscriptions: make(map[string]bool),
+	}
+
+	for _, topic := range topics {
+		b.subscriptions[topic] = true
+	}
+
+	ctx := context.Background()
+	b.ctx, b.cancel = context.WithCancel(ctx)
+
+	b.client = redis.NewClient(&redis.Options{
+		Addr: getRedisHost(),
+	})
+
+	go b.lc.WatchContext(ctx)
+	go b.run()
+
+	return b
+}
+
+func (b *subscriber) run() {
+	defer func() {
+		b.cancel()
+		_ = b.client.Close()
+
+		b.lc.ShutdownCompleted()
+	}()
+
+	var subs *redis.PubSub
+
+	for topic := range b.subscriptions {
+		if subs == nil {
```

Comment:
The control flow inside this loop is weird & their is an empty `if err != nil` down below. What is going on in this?